etcd/v3/kv_etcd.go

```diff
@@ -580,14 +543,37 @@ func (et *etcdKV) CompareAndDelete(
 ) (*kvdb.KVPair, error) {
 	key := et.domain + kvp.Key
 	ctx, cancel := et.Context()
+
+	cmp := e.Compare(e.Value(key), "=", string(kvp.Value))
 	if (flags & kvdb.KVModifiedIndex) != 0 {
+		cmp = e.Compare(e.ModRevision(key), "=", int64(kvp.ModifiedIndex))
+	}
+	for i := 0; i < timeoutMaxRetry; i++ {
 		txnResponse, txnErr := et.kvClient.Txn(ctx).
-			If(e.Compare(e.ModRevision(key), "=", int64(kvp.ModifiedIndex))).
+			If(cmp).
 			Then(e.OpDelete(key)).
 			Commit()
 		cancel()
 		if txnErr != nil {
-			return nil, txnErr
+			// Check if we need to retry
+			retry, txnErr := isRetryNeeded(txnErr, "cad", key, i)
```

Comment:
key not found here == error ? 
Since we are deleting should we really care for key not found ?

Reply:
isRetryNeeded will return ErrNotFound immediately and down below we handle that and return success.

Reply:
isRetryNeeded will return retry=false, which will caused !retry (line 560) to be true and return txnErr ?

Reply:
ah! good catch. Will fix this.